README.md
# Weblate prometheus exporter

[![weblate prometheus exporter](https://github.com/loktionovam/weblate_exporter/actions/workflows/ci.yml/badge.svg)](https://github.com/loktionovam/weblate_exporter/actions/workflows/ci.yml)

[![codecov](https://codecov.io/gh/loktionovam/weblate_exporter/branch/main/graph/badge.svg?token=3OWCAKRWEA)](https://codecov.io/gh/loktionovam/weblate_exporter)

Prometheus exporter for Weblate server metrics ([Weblate metrics described here](https://docs.weblate.org/en/latest/api.html))

Supported versions:

* Weblate >= 4.9

Basic metrics
| Metric                                      | Type  | Description                                                                                                                         | Labels                                                 |
|---------------------------------------------|-------|-------------------------------------------------------------------------------------------------------------------------------------|--------------------------------------------------------|
| `weblate_exporter_app_up`                   | Gauge | Weblate application is running up                                                                                                   | `name`, `weblate_api_url`                              |
| `weblate_exporter_app_units`                | Gauge | Total number of [units](https://docs.weblate.org/en/latest/api.html?highlight=units#units)                                          | `name`, `weblate_api_url`                              |
| `weblate_exporter_app_units_translated`     | Gauge | Number of translated [units](https://docs.weblate.org/en/latest/api.html?highlight=units#units)                                     | `name`, `weblate_api_url`                              |
| `weblate_exporter_app_users`                | Gauge | Number of weblate users                                                                                                             | `name`, `weblate_api_url`                              |
| `weblate_exporter_app_changes`              | Gauge | Number of [translation changes](https://docs.weblate.org/en/latest/api.html?highlight=changes#get--api-changes-)                    | `name`, `weblate_api_url`                              |
| `weblate_exporter_app_projects`             | Gauge | Number of [projects](https://docs.weblate.org/en/latest/user/translating.html#translation-projects)                                 | `name`, `weblate_api_url`                              |
| `weblate_exporter_app_components`           | Gauge | Number of [components](https://docs.weblate.org/en/latest/admin/projects.html#component)                                            | `name`, `weblate_api_url`                              |
| `weblate_exporter_app_translations`         | Gauge | Number of [translations](https://docs.weblate.org/en/latest/api.html?highlight=units#translations)                                  | `name`, `weblate_api_url`                              |
| `weblate_exporter_app_languages`            | Gauge | Number of [used languages](https://docs.weblate.org/en/latest/api.html?highlight=units#languages)                                   | `name`, `weblate_api_url`                              |
| `weblate_exporter_app_configuration_errors` | Gauge | Number of [configuration errors](https://docs.weblate.org/en/latest/api.html?highlight=units#metrics)                               | `name`, `weblate_api_url`                              |
| `weblate_exporter_app_suggestions`          | Gauge | Number of pending [suggestions](https://docs.weblate.org/en/latest/api.html?highlight=units#metrics)                                | `name`, `weblate_api_url`                              |
| `weblate_exporter_app_celery_queues`        | Gauge | Lengths of various celery queues, see [Background tasks using Celery](https://docs.weblate.org/en/latest/admin/install.html#celery) | `name`, `weblate_api_url`, `weblate_celery_queue_name` |


## Building and running

### Prerequisites

* docker engine >= 20.10
* helm >= 3.7.1
* helm-docs >= 1.5.0
* make
* python >= 3.10 (used by tests only)
* python3-venv (used by tests only)

### Setup an environment for developing and testing

* Install prerequisites for developing and testing:

  ```shell
  sudo apt-get install python3-venv
  python3 -m venv venv
  source venv/bin/activate
  python -m pip install --upgrade pip

  # Install for tests only
  pip install -r requirements-test.txt

  # Install for developing
  pip install -r requirements-dev.txt
  ```

### Build and test a docker image

* Set up an image name and tag (optional):

    ```shell
    export WEBLATE_EXPORTER_IMAGE_NAME=yourname/weblate_exporter
    export WEBLATE_EXPORTER_IMAGE_TAG=0.1.0
    ```

* Build and test the image:

    ```shell
    make build-images
    make test-images
    ```

### Run the weblate_exporter

#### Helm

* Add weblate-exporter helm repository:

  ```shell
    helm repo add weblate-exporter https://raw.githubusercontent.com/loktionovam/weblate_exporter/gh-pages/
    helm repo update
  ```

* Search available weblate-exporter helm charts:

  ```shell
    helm search repo --versions weblate-exporter

    NAME                             	CHART VERSION	APP VERSION	DESCRIPTION
    weblate-exporter/weblate-exporter	0.3.2        	v0.3.2     	Weblate prometheus metrics exporter

  ```

* Deploy new helm release:

  ```shell
  helm install weblate-exporter weblate-exporter/weblate-exporter \
    --version 0.3.2 \
    --set config.weblateAPIUrl="http://weblate.example.com:8080/api/" \
    --set config.weblateAPIKey="secret_api_key"

    helm test weblate-exporter
  ```

#### Docker

* Running using environment variables:

  ```shell
    docker run -p 9867:9867  -d -e WEBLATE_API_URL=http://weblate.example.com/api/ -e WEBLATE_API_KEY=secret_api_key yourname/weblate_exporter:0.1.0

    # Get the metrics
    curl http://localhost:9867/metrics/
  ```

* Running using a configuration file:

  ```shell
    mv config.yaml.example weblate_exporter.yaml

    # Setup weblate_api_url and weblate_api_key variables in weblate_exporter.yaml

    cat weblate_exporter.yaml                                                                                                                                                                                                              ─╯
    ---
    weblate_api_url: http://weblate.example.com/api/
    weblate_api_key: secret_api_key
    weblate_exporter_bind_port: 9867

    # Run weblate_exporter docker container:
    docker run -p 9867:9867  -d -v $(pwd)/weblate_exporter.yaml:/etc/weblate_exporter.yaml  yourname/weblate_exporter:0.1.0

    # Get the metrics
    curl http://localhost:9867/metrics/

  ```

### Grafana

How to import [Weblate metrics dashboard](./grafana/dashboards/weblate.json) described [here](https://grafana.com/docs/grafana/latest/dashboards/export-import/#import-dashboard)

![Grafana Dashboard (Summary)](grafana/img/dashboard-summary.png)
![Grafana Dashboard (Graphs)](grafana/img/dashboard-graphs.png)

## Developing and testing weblate exporter

* Install prerequisites [as described here](#setup-an-environment-for-developing-and-testing) and activate python virtual environment
* Install pre-commit hooks

  ```shell
  source venv/bin/activate
  pre-commit install
  pre-commit install-hooks
  ```

* Format the code and run tests:

  ```shell
  make fmt
  make test-apps
  ```


CHANGELOG.md
# Changelog


## v1.1.2 (2022-01-18)

### Other

* Release v1.1.2. [Aleksandr Loktionov]

* Merge pull request #13 from loktionovam/bugfix/iterator. [Aleksandr Loktionov]

  Upgrade docker image due to vulnerable expat package

* Upgrade docker image due to vulnerable expat package. [Aleksandr Loktionov]

  * Upgrade expat package
  * Add iterator to WeblateMetrics class

* Release v1.1.1. [Aleksandr Loktionov]


## v1.1.1 (2021-12-25)

### Other

* Release changelog v1.1.1. [Aleksandr Loktionov]

* Merge pull request #12 from loktionovam/bugfix/remove-dict-as-default-val. [Aleksandr Loktionov]

  Remove default empty dicts from WeblateMetrics constructor

* Remove default empty dicts from WeblateMetrics constructor. [Aleksandr Loktionov]


## v1.1.0 (2021-11-27)

### Other

* Release changelog v1.1.0. [Aleksandr Loktionov]

* Merge pull request #11 from loktionovam/feature/changelog-autogeneration. [Aleksandr Loktionov]

  Add changelog generation

* Add changelog generation. [Aleksandr Loktionov]

* Merge pull request #10 from loktionovam/feature/grafana-dashboards. [Aleksandr Loktionov]

  Add grafana dashboards

* Add grafana dashboards. [Aleksandr Loktionov]

  * Replace multiple celery metrics by single one
  * Rename weblate_exporter_port to weblate_exporter_bind_port due to conflict with k8s env variables injected to pod
  * Add servicemonitor to helm chart


## v0.3.4 (2021-11-26)

### Other

* Merge pull request #9 from loktionovam/docs/helm-readme. [Aleksandr Loktionov]

  Add helm chart readme autogeneration

* Add helm chart readme autogeneration. [Aleksandr Loktionov]


## v0.3.3 (2021-11-25)

### Other

* Merge pull request #8 from loktionovam/tests/helm-chart-tests. [Aleksandr Loktionov]

  Add tests for helm chart

* Add tests for helm chart. [Aleksandr Loktionov]


## v0.3.2 (2021-11-25)

### Other

* Merge pull request #7 from loktionovam/feature/publish-helm-chart. [Aleksandr Loktionov]

  Publish helm chart

* Publish helm chart. [Aleksandr Loktionov]


## v0.3.1 (2021-11-25)

### Other

* Merge pull request #6 from loktionovam/bugfix/patch-helm-values. [Aleksandr Loktionov]

  Patch helm values accordingly to docker image version

* Patch helm values accordingly to docker image version. [Aleksandr Loktionov]


## v0.3.0 (2021-11-24)

### Other

* Merge pull request #5 from loktionovam/feature/helm-chart. [Aleksandr Loktionov]

  Add weblate exporter helm chart

* Add weblate exporter helm chart. [Aleksandr Loktionov]


## v0.2.0 (2021-11-24)

### Other

* Merge pull request #4 from loktionovam/feature/push-images-to-registry. [Aleksandr Loktionov]

  Publish weblate exporter docker image

* Publish weblate exporter docker image. [Aleksandr Loktionov]

* Merge pull request #3 from loktionovam/bugfix/metrics-duplication. [Aleksandr Loktionov]

  Fix metrics duplication

* Fix metrics duplication. [Aleksandr Loktionov]

* Merge pull request #2 from loktionovam/docs/initital-documentation. [Aleksandr Loktionov]

  Add LICENSE, README.md and docstrings

* Add LICENSE, README.md, docstrings. [Aleksandr Loktionov]

* Merge pull request #1 from loktionovam/test/sometest. [Aleksandr Loktionov]

  Upload pytest report to codecov

* Upload pytest report to codecov. [Aleksandr Loktionov]

* Initial commit. [Aleksandr Loktionov]


## v0.1.0 (2021-11-20)

### Other

* Initial commit. [Aleksandr Loktionov]




tests/apps/test_cli.py
from weblate_exporter.cli import parse_args, run_exporter
from unittest import TestCase, mock
from multiprocessing import Process
from tests.apps.test_collector import METRICS, API_URL
import os
import responses
import requests
from time import sleep

DATA_DIR = os.path.join(os.path.dirname(os.path.abspath(__file__)), "data")


class TestCLI(TestCase):
    def setUp(self):
        responses.mock.start()
        responses.add(responses.GET, API_URL + "metrics/", json=METRICS, status=200)
        responses.add_passthru("http://127.0.0.1:9867/metrics")

    def tearDown(self):
        """Disable responses."""
        responses.mock.stop()
        responses.mock.reset()

    def test_parse_args_config(self):
        config_path = os.path.join(DATA_DIR, "config.yaml")
        args = parse_args(["--config", config_path])
        assert args.weblate_exporter_bind_port == 9867
        assert args.weblate_api_url == "http://test-weblate:8080/api/"
        assert args.weblate_api_key == "test_weblate_api_key"

    @mock.patch.dict(
        os.environ,
        {
            "WEBLATE_API_URL": "http://env-test-weblate:8080/api/",
            "WEBLATE_API_KEY": "env_test_weblate_api_key",
            "WEBLATE_EXPORTER_BIND_PORT": "9868",
        },
    )
    def test_parse_args_env(self):
        args = parse_args([])

        assert args.weblate_exporter_bind_port == 9868
        assert args.weblate_api_url == "http://env-test-weblate:8080/api/"
        assert args.weblate_api_key == "env_test_weblate_api_key"

    def test_run_exporter(self):
        args = parse_args(["--weblate_api_url", API_URL])

        p = Process(target=run_exporter, args=(args,))
        try:
            p.start()
            sleep(1)
            response = requests.get("http://127.0.0.1:9867/metrics")
            self.assertEqual(200, response.status_code)
        finally:
            p.terminate()
            p.join()


tests/images/conftest.py
import pytest
import subprocess
import os
import testinfra

WEBLATE_EXPORTER_IMAGE_NAME = os.environ.get(
    "WEBLATE_EXPORTER_IMAGE_NAME", "loktionovam/weblate_exporter"
)
WEBLATE_EXPORTER_IMAGE_TAG = os.environ.get("WEBLATE_EXPORTER_IMAGE_TAG", "dev")


# scope='session' uses the same container for all the tests;
# scope='function' uses a new container per test function.
@pytest.fixture(scope="class")
def host(request):
    # run a container
    docker_id = (
        subprocess.check_output(
            [
                "docker",
                "run",
                "-d",
                f"{WEBLATE_EXPORTER_IMAGE_NAME}:{WEBLATE_EXPORTER_IMAGE_TAG}",
            ]
        )
        .decode()
        .strip()
    )
    # return a testinfra connection to the container
    host = testinfra.get_host("docker://" + docker_id)
    request.cls.host = host
    yield host
    # at the end of the test suite, destroy the container
    subprocess.check_call(["docker", "rm", "-f", docker_id])


tests/apps/test_logger.py
from unittest import TestCase, mock
import os
import importlib
import weblate_exporter.logger


class TestLogger(TestCase):
    def test_log(self):
        importlib.reload(weblate_exporter.logger)
        log = weblate_exporter.logger.log
        self.assertEqual(20, log.level)

        with self.assertLogs(logger=log, level=log.level) as cm:
            log.info("info message")
            log.warning("warn message")
        self.assertEqual(
            cm.output,
            [
                "INFO:weblate_exporter:info message",
                "WARNING:weblate_exporter:warn message",
            ],
        )

    @mock.patch.dict(os.environ, {"DEBUG": "1"})
    def test_log_debug(self):
        importlib.reload(weblate_exporter.logger)
        log = weblate_exporter.logger.log
        self.assertEqual(10, log.level)

        with self.assertLogs(logger=log, level=log.level) as cm:
            log.info("info message")
            log.warning("warn message")
            log.debug("debug message")
        self.assertEqual(
            cm.output,
            [
                "INFO:weblate_exporter:info message",
                "WARNING:weblate_exporter:warn message",
                "DEBUG:weblate_exporter:debug message",
            ],
        )


helm/release_helm_chart.py
#!/usr/bin/env python3

import collections.abc
import os
import subprocess
import ruamel.yaml

"""
The script patches Chart.yaml, values.yaml by current version of an application
and creates a helm chart package.
"""

script_path = os.path.dirname(os.path.abspath(__file__))
chart_path = os.path.join(script_path, "charts/weblate-exporter")
chart_info_path = os.path.join(chart_path, "Chart.yaml")

chart_values_path = os.path.join(chart_path, "values.yaml")

app_version = os.environ.get("WEBLATE_EXPORTER_IMAGE_TAG", "v0.1.0")
app_image = os.environ.get("WEBLATE_EXPORTER_IMAGE_NAME", "loktionovam/weblate_exporter")
chart_version = app_version.strip("v")


# Update nested dict based on https://stackoverflow.com/a/3233356
def update_dict(src_dict: dict = {}, patch_dict: dict = {}) -> dict:
    for k, v in patch_dict.items():
        if isinstance(v, collections.abc.Mapping):
            src_dict[k] = update_dict(src_dict.get(k, {}), v)
        else:
            src_dict[k] = v
    return src_dict


def patch_yaml_file(yaml_file_path: str = "", patch_dict: dict = {}) -> None:
    print(f"Apply patch: {patch_dict} to {yaml_file_path}")
    yaml = ruamel.yaml.YAML()
    yaml.preserve_quotes = True
    with open(yaml_file_path) as yaml_fh:
        yaml_values = yaml.load(yaml_fh)

    patched_yaml_values = update_dict(yaml_values, patch_dict)
    with open(yaml_file_path, "w") as yaml_fh:
        yaml.dump(patched_yaml_values, yaml_fh)


patch_yaml_file(
    yaml_file_path=chart_info_path,
    patch_dict={
        "version": chart_version,
        "appVersion": app_version
    })

patch_yaml_file(
    yaml_file_path=chart_values_path,
    patch_dict={
        "image": {
            "repository": app_image,
            "tag": app_version,
        }
    })


# helm_package_output = subprocess.check_output(
#     [
#         "helm",
#         "package",
#         chart_path,
#     ]
# ).decode().strip()
#
# print(helm_package_output)


weblate_exporter/collector.py
from wlc import Weblate, WeblateException
from prometheus_client.core import GaugeMetricFamily
from weblate_exporter.logger import log
from requests.exceptions import ConnectionError


class WeblateMetrics:
    """
    The data class which parses raw metrics from Weblate described here:

    https://docs.weblate.org/en/latest/admin/install.html?highlight=metrics#monitoring-weblate

    and represents them in a prometheus format.

    Example of metrics:

        weblate_exporter_app_units
        weblate_exporter_app_units_translated
        ....
        ....

    Celery queues stored as weblate_exporter_app_celery_queues metric
    with an additional label weblate_celery_queue_name.

    """

    def __init__(self, raw_metrics: dict, additional_labels: dict):
        self.metrics_prefix = "weblate_exporter_app"
        labels_names = ["name"]
        labels_values = [raw_metrics["name"]]

        for k, v in additional_labels.items():
            labels_names.append(k)
            labels_values.append(v)

        self.data = {
            f"{self.metrics_prefix}_up": GaugeMetricFamily(
                f"{self.metrics_prefix}_up",
                "weblate application is up",
                labels=["name"],
            ),
        }

        for metric, value in raw_metrics.items():
            if metric != "name":
                metric_name = f"{self.metrics_prefix}_{metric}"
                if metric == "celery_queues":
                    self.data[metric_name] = GaugeMetricFamily(
                        metric_name,
                        "",
                        labels=labels_names + ["weblate_celery_queue_name"],
                    )

                    for queue_label, queue_value in raw_metrics[metric].items():
                        self.data[metric_name].add_metric(
                            labels_values + [queue_label], queue_value
                        )
                else:
                    self.data[metric_name] = GaugeMetricFamily(
                        metric_name,
                        "",
                        labels=labels_names,
                    )
                    self.data[metric_name].add_metric(labels_values, value)

    def __iter__(self):
        for metric in self.data.values():
            yield metric


class WeblateCollector:
    """
    The custom prometheus collector which fetches metrics data from Weblate instance and
    store them in prometheus format. Instance of the class is used by a prometheus client registry.

    https://github.com/prometheus/client_python#custom-collectors
    """

    def __init__(
        self,
        api_url: str,
        api_key: str = None,
    ):
        """
        :param api_url: Weblate API root endpoint, i.e. http://weblate:8080/api/
                        Notice, that you should use '/' as a trailing slash.

                    More information about Weblate API here: https://docs.weblate.org/en/latest/api.html
        :param api_key: authorization token which described here https://docs.weblate.org/en/latest/api.html?highlight=token#authentication-and-generic-parameters
        """

        self.api_key = api_key
        self.api_url = api_url
        self._raw_metrics = {}
        self._wlc = Weblate(key=api_key, url=api_url)
        self.metrics = None

    def _fetch_raw_metrics(self) -> None:
        try:
            log.debug(f"Try to fetch metrics from: {self.api_url}")
            self._raw_metrics = self._wlc.get("metrics/")
            self._raw_metrics["up"] = 1
        except (WeblateException, ConnectionError) as error:
            log.warning(f"Can't fetch metrics from weblate: {error}")
            self._raw_metrics = {"up": 0, "name": "unknown"}
        finally:
            log.debug(f"Fetched raw metrics: {self._raw_metrics}")

    def collect(self):
        self._fetch_raw_metrics()
        self.metrics = WeblateMetrics(
            self._raw_metrics, additional_labels={"weblate_api_url": self.api_url}
        )
        for metric in self.metrics:
            yield metric


weblate_exporter/cli.py
import os
import sys
import time
import configargparse
from prometheus_client import start_http_server
from prometheus_client.core import REGISTRY

from weblate_exporter.collector import WeblateCollector
from weblate_exporter.logger import log


def parse_args(args):
    parser = configargparse.ArgParser(
        description="Weblate exporter args weblate address, api key and weblate exporter port",
        default_config_files=["/etc/weblate_exporter.yaml"],
    )

    parser.add("-c", "--config", help="config filename", is_config_file=True)

    parser.add(
        "-s",
        "--weblate_api_url",
        metavar="weblate_api_url",
        required=False,
        help="Weblate API URL",
        default=os.environ.get("WEBLATE_API_URL", "http://weblate:8080/api/"),
    )
    parser.add(
        "--weblate_api_key",
        metavar="weblate_api_key",
        required=False,
        help="weblate api key",
        default=os.environ.get("WEBLATE_API_KEY"),
    )
    parser.add(
        "-p",
        "--weblate_exporter_bind_port",
        metavar="weblate_exporter_bind_port",
        required=False,
        type=int,
        help="Listen to this port",
        default=int(os.environ.get("WEBLATE_EXPORTER_BIND_PORT", "9867")),
    )

    return parser.parse_args(args)


def run_exporter(args):
    log.info(f"Start weblate exporter for: {args.weblate_api_url}")
    collector = WeblateCollector(
        api_url=args.weblate_api_url,
        api_key=args.weblate_api_key,
    )
    REGISTRY.register(collector)

    start_http_server(args.weblate_exporter_bind_port)
    log.info(f"Serving at port: {args.weblate_exporter_bind_port}")
    while True:
        time.sleep(1)


def main():

    args = parse_args(sys.argv[1:])
    run_exporter(args)


if __name__ == "__main__":
    main()
